climpred/prediction.py

```diff
@@ -257,11 +260,22 @@ def compute_hindcast(
     for i in forecast.lead.values:
         if max_dof:
             forecast, reference = reduce_time_series(forecast, reference, i)
-        # take lead year i timeseries and convert to real time
+
+        # take lead  i timeseries and convert to real time based on temporal
+        # resolution of lead (lead_res)
+        offset_args_dict = get_lead_pdoffset_args(forecast['lead'].attrs['units'], i)
         a = forecast.sel(lead=i).drop('lead')
-        a['time'] = [int(t + i) for t in a.time.values]
-        # take real time reference of real time forecast years
+        a['time'] = pd.to_datetime(
+            a['time'].dt.strftime('%Y%m%d 00:00')
```

Comment:
Wondering if it's better to have `%Y%m%d %H:00` because I think some models start running at 12Z

Reply:
I think we should come up with a common set of expectations about dims and specifically `init` and `lead` and `time`.
- which hour, secs, ... do we use internally when this is not specified?
- do we accept `int` representing years for `init` and `lead` and `time` (still strongly in favour for decadal predictions), but internally this can all be changed to `datetime`
-

Reply:
> do we accept int representing years for init and lead and time (still strongly in favour for decadal predictions), but internally this can all be changed to datetime

I think this is a good compromise. To just accept `int` and do operations on that would balloon the code a bit. I think accepting `int` and then converting it to a standard datetime is good.

Reply:
@bradyrx do you remember the thread when we discussed he conventions for lead init and time the last time? Couldn’t find it. This might be a good start

Reply:
I like the idea of dealing with the lead timestep via `ds.lead.attrs`, although maybe we could even write a function that guesses this from the increment in the lead dimension?

Reply:
@ahuang11 I'm not sure how best to handle the issue of including the hours in the date time information.  In order to find matching times between model and verification, the datetime has to match exactly.  Typically for daily data, the verification datasets will not specify the hour or if they do, it is an arbitrary choice with no real meaning. For monthly data, the same issue arises for including the day of the month.  The verification data will have only the month and no day or an arbitrary choice of day (e.g. 01 or 15).   Any ideas for how best to handle this?

Reply:
I haven't tried out the branch, but one suggestion is:

```
import pandas as pd
import warnings

dts1 = pd.date_range('2017-02-01 12:00', '2017-03-01 12:00')
dts2 = pd.date_range('2017-02-01 00:00', '2017-03-01 00:00')

if len(set(dts1) & set(dts2)) == 0:
    timedelta_offset = pd.Timedelta((dts1 - dts2).values.mean())
    dts2 += timedelta_offset
    warnings.warn(f'dts2 has been offset by {timedelta_offset}')
```

Or have something similar to pd.to_numeric() keyword:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Where ignore would just replace hours with 0 (your current method), but either way, it's a big step forward already to support subannual predictions so we can add this in another PR

Reply:
Further question about accepting integers -- if we do accept integers, then we have to make some sort of assumption about what that integer means.  Do we only accept integers with lead.units=annual and assume the integers in this case are years?  Do we try to interpret any integer?  I could imagine someone trying integers like 19820101, which is probably not what we intend.

Reply:
> I like the idea of dealing with the lead timestep via `ds.lead.attrs`, although maybe we could even write a function that guesses this from the increment in the lead dimension?
@bradyrx I think using units is a good choice here.  Typically in a netcdf file, all dimensions should have units, but there is currently no formal definition for units of the lead dimension and it is typically just ignored for the lead dimension (at least for subseasonal and seasonal data.  Not sure about decadal).  Its about time someone defined it, so I think we can be the ones to do that. 
@aaronspring I could not think of any robust way to guess the lead timestep based on the way the DPLE, SubX, or NMME databases are setup. I think it will have to be defined by the user.  I should be able to get SubX and NMME database to add this to the metadata in the repository.

Reply:
Currently climpred takes integers for Leads and inits. The init int means initialisation at 01-jan-int 00:00 and lead=1 means the annual average for that year. Lead=2 would be int+1...

This time model of climpred works for annual timestepping but we obviously need torethink this now.

Reply:
@bradyrx @ahuang11 and me had discussion on this which we never finished. https://github.com/bradyrx/climpred/pull/172

Reply:
I think it’s less important how we deal with the current int format. Let’s rather decide on a clever way for datetime init and lead and then write a wrapper for the currently used int datasets.

Reply:
I am testing code to check for init types of integer (for annual data) and then convert to pd.DatetimeIndex.  In the testing for integers, I am using the quickstart tutorial data. That data appears to be stored as int64 in the netcdf file (according to ncdump).  Once it is read in (hind = climpred.tutorial.load_dataset('CESM-DP-SST')), it becomes type float64 (see below).  I was expecting and testing for integers.  Is this what we should expect and handle for annual data or should we be expecting integers?

Float64Index([1954.0, 1955.0, 1956.0, 1957.0, 1958.0, 1959.0, 1960.0, 1961.0,
              1962.0, 1963.0, 1964.0, 1965.0, 1966.0, 1967.0, 1968.0, 1969.0,
              1970.0, 1971.0, 1972.0, 1973.0, 1974.0, 1975.0, 1976.0, 1977.0,
              1978.0, 1979.0, 1980.0, 1981.0, 1982.0, 1983.0, 1984.0, 1985.0,
              1986.0, 1987.0, 1988.0, 1989.0, 1990.0, 1991.0, 1992.0, 1993.0,
              1994.0, 1995.0, 1996.0, 1997.0, 1998.0, 1999.0, 2000.0, 2001.0,
              2002.0, 2003.0, 2004.0, 2005.0, 2006.0, 2007.0, 2008.0, 2009.0,
              2010.0, 2011.0, 2012.0, 2013.0, 2014.0, 2015.0, 2016.0, 2017.0],
             dtype='float64', name='init')

Reply:
Sorry for getting back to you so late. I didn't see this comment.

So I'm seeing from `ncdump` that the `init` values are stored as floats:
<img width="404" alt="Screen Shot 2019-12-12 at 2 27 09 PM" src="https://user-images.githubusercontent.com/8881170/70750348-7f7cd080-1ceb-11ea-9a99-899d9b54330b.png">

So when it's pulled into `xarray` it's stored as a `Float64Index`:
```python
hind = climpred.tutorial.load_dataset('CESM-DP-SST')
hind['init'].to_index()
>>> Float64Index([1954.0, 1955.0, 1956.0, 1957.0, 1958.0, 1959.0, 1960.0, 1961.0,
              1962.0, 1963.0, 1964.0, 1965.0, 1966.0, 1967.0, 1968.0, 1969.0,
              1970.0, 1971.0, 1972.0, 1973.0, 1974.0, 1975.0, 1976.0, 1977.0,
              1978.0, 1979.0, 1980.0, 1981.0, 1982.0, 1983.0, 1984.0, 1985.0,
              1986.0, 1987.0, 1988.0, 1989.0, 1990.0, 1991.0, 1992.0, 1993.0,
              1994.0, 1995.0, 1996.0, 1997.0, 1998.0, 1999.0, 2000.0, 2001.0,
              2002.0, 2003.0, 2004.0, 2005.0, 2006.0, 2007.0, 2008.0, 2009.0,
              2010.0, 2011.0, 2012.0, 2013.0, 2014.0, 2015.0, 2016.0, 2017.0],
             dtype='float64', name='init')
```

There is also a type `Int64Index`:
```python
import numpy as np
import xarray as xr
test_ds = xr.DataArray(np.random.rand(10), dims=['time'], coords={'time': np.arange(10)})
test_ds['time'].to_index()
>>> Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype='int64', name='time')
```

`xarray` pulls from the `pandas` numeric indices here: https://pandas.pydata.org/pandas-docs/stable/reference/indexing.html. So it looks like if you just explicitly test for `Float64Index` and `Int64Index` you should have it covered. Then you could convert those over to `pandas` datetime. You'd test in the same way I suggested testing for the cftime/datetime indices.